### PR TITLE
[HTTPD] support for multiple simultaneous requests (try 3)  (IDFGH-9868)

### PR DESCRIPTION
**Pull request by chipweinberger**
**Related Issue:** https://github.com/espressif/esp-idf/issues/10594

**Previous PR:** https://github.com/espressif/esp-idf/pull/10669

Add basic support for supporting multiple simultaneous requests in HTTPD server.

**This is the 3rd try of this PR:**
- **1st**: worked well
- **2nd**: tried to simplified the user api, but made things too complicated for the server code
- **3rd**: went back to the first approach, with slightly cleaner api

**Note**: This PR does _not_ add _native_ support for simultaneous requests, but instead makes it more ergonomic for users to move a request to a thread that they must create themselves.

**Advantages over raw sockets**
- **easier reply:** Users can use all the existing httpd functions (i.e. `httpd_resp_sendstr`) to send their response instead of raw socket send()
- **LRU safety:** A long running async socket will not get closed from underneath us if we run out of sockets

**Implementation:**

- an request handler will first call `httpd_req_async_handler_begin()`
  - this gives the user exclusive use of a new request object (to not interfere with the http server's request)
  - this will mark the socket as "in use" to guarentee the socket to not be LRU purged
- async handler must call `httpd_req_async_handler_complete()` when done 
  - this frees our request object.
  - this will mark the socket as "no longer in use" to allow it to be closed (LRU purged)

**Example**

- add an example to ESP-IDF, showing usage.

### CODE REVIEW

**Comment by chipweinberger** [comment]
@hmalpani

File: examples/protocols/http_server/async_handlers/main/main.c [hunk 1]
```diff
@@ -0,0 +1,352 @@
...
+#include "esp_netif.h"
+#include "esp_eth.h"
+#include "protocol_examples_common.h"
+#include "esp_tls_crypto.h"
+#include <esp_http_server.h>
+
+/* An example that demonstrates multiple
+   long running http requests running in parallel.
+
+   In this example, multuple long http request can run at
```
Comment (KaeLL) [inline examples/protocols/http_server/async_handlers/main/main.c:27]: typo

**Review by hmalpani — COMMENTED** [review]
Hello @chipweinberger 
This approach looks simpler to me.
Although when I tried these changes, I got into some errors. The program crashed when calling `long` handler. Can you please check and update the PR accordingly?
Thanks!

**Comment by chipweinberger** [comment]
**It works now.**

I introduced a small bug right before pushing. 🤦 

in `httpd_req_async_handler_begin` I was calling a memcpy out of order, overwriting a previous assignment.

bad code:
```
async->aux = malloc(sizeof(struct httpd_req_aux));
memcpy(async, r, sizeof(httpd_req_t)); // overwrites previous assignment.
memcpy(async->aux, r->aux, sizeof(struct httpd_req_aux));
```

I just had to move the 1st memcpy first before `async->aux =`.

File: examples/protocols/http_server/async_handlers/main/main.c [hunk 2]
```diff
@@ -0,0 +1,352 @@
+/* Async Request Handlers HTTP Server Example
+
+   This example code is in the Public Domain (or CC0 licensed, at your option.)
+
+   Unless required by applicable law or agreed to in writing, this
+   software is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+   CONDITIONS OF ANY KIND, either express or implied.
+*/
+#include "freertos/freertos.h"
```
[the remark this answers is not in the document]
Comment (chipweinberger) [reply examples/protocols/http_server/async_handlers/main/main.c:9]: got it.

**Comment by hmalpani** [comment]
sha=1289d69870dc5faee84756e168dbacdb59f62a4a

**Comment by hmalpani** [comment]
Hello @chipweinberger 
Can you please sign the CLA agreement. Thanks!

**Comment by chipweinberger** [comment]
signed

**Comment by chipweinberger** [comment]
any updates? thanks

**Comment by hmalpani** [comment]
Hello @chipweinberger 
I have created an internal MR which is under review. It will get reflected on GitHub once it gets merged.
Thanks!

**Comment by chipweinberger** [comment]
perhaps I can split the PR up into 2 to make it be reviewed faster? Let me know if that would be better.

1. core changes (which is only a small diff in 4 files)
2. new example

**Comment by hmalpani** [comment]
> perhaps I can split the PR up into 2 to make it reviewed faster? Let me know if that would be better.
> 
>     1. core changes (which is only a small diff in 4 files)
> 
>     2. new example

Hello @chipweinberger 
The PR is already in review and should not take long before getting merged. So no need to update the PR.
Thanks!

**Comment by hmalpani** [comment]
PR has been merged into master via commit: 3824eba04dbdc1a962b9a039bbbb8df87436f682

Closing this PR.

**Comment by ToonVanEyck** [comment]
I have implemented what's shown in the example and it seems to be working. But when doing the same with POST requests, it crashes. 

In my POST request I was trying to read a chuck of data and process it before reading the next chunk. Would this be possible?

The end goal is to keep streaming log data through a we socket. But currently the stream stops when doing the old synchronous POST request and I was hoping async would solve that.

**Pull request closed without merging**